tools/ci/restrict_file_types.py

```diff
@@ -0,0 +1,67 @@
+import glob
+import re
+import os
+import sys
+import time
+from collections import namedtuple
+
+Failure = namedtuple("Failure", ["filename", "lineno", "message"])
+
+RED = "\033[0;31m"
+GREEN = "\033[0;32m"
+BLUE = "\033[0;34m"
+NC = "\033[0m"  # No Color
+
+def print_error(message: str, filename: str, line_number: int):
+    if os.getenv("GITHUB_ACTIONS") == "true": # We're on github, output in a special format.
+        print(f"::error file={filename},line={line_number},title=Restricted Type in File::{filename}:{line_number}: {RED}{message}{NC}")
+    else:
+        print(f"{filename}:{line_number}: {RED}{message}{NC}")
+
+if __name__ == "__main__":
+    print("restrict_file_types started")
+
+    exit_code = 0
+    start = time.time()
+
+    dm_files = glob.glob("**/*.dm", recursive=True)
+
+    if len(sys.argv) > 1:
+        dm_files = [sys.argv[1]]
+
+    all_failures = []
+
+    for code_filepath in dm_files:
+        with open(code_filepath, encoding="UTF-8") as code:
+            filename = code_filepath.split(os.path.sep)[-1]
+
+            restrict_regex = re.match(r"RESTRICT_TYPE\((.+)\)", code.read())
+            if(not restrict_regex):
+                continue
+
+            type_path = restrict_regex.group(1)
+
+            # These regexes could probably usee some refinement, but they're the best I could come up with.
+            proc_search = re.compile(r'^(/[\w/]{3,}?)\/(?:proc\/)?\w+\(') # Search for any procs that are not of this type
+            def_search = re.compile(r'(^(?:/(?:\w+))+)') # Search for any definition that are not of this type
+            code.seek(0)
+            for idx, line in enumerate(code):
+                if(rematch_result := re.search(proc_search, line)):
+                    rematch_result = rematch_result.group(1)
+                    if(type_path != rematch_result):
+                        all_failures += [Failure(code_filepath, idx + 1, f"'{rematch_result}' proc found in a file restricted to '{type_path}'")]
+                elif(rematch_result := re.search(def_search, line)):
+                    rematch_result = rematch_result.group(1)
+                    if(type_path != rematch_result):
+                        all_failures += [Failure(code_filepath, idx + 1, f"'{rematch_result}' type definition found in a file restricted to '{type_path}'")]
```

Comment:
Here's a version of the regex that handles both cases, and avoids some potential issues like start-of-line comments:
```suggestion
            # Matches a definition into four groups:
            # 1: The typepath or /proc (for global procs), required. First character is handled specially, to avoid picking up start-of-line comments.
            # 2: /proc, if present and not the first segment.
            # 3: /a_proc( and everything after it, if present. Probably not useful by itself
            # 4: /a_proc from #3, if present.
            definition_matcher = re.compile(r'^(/[\w][\w/]*?)(/proc)?((/[\w]+)\(.*)?$')
            code.seek(0)
            for idx, line in enumerate(code):
                if(rematch_result := re.search(definition_matcher, line)):
                    rematch_result = rematch_result.group(1)
                    if(type_path != rematch_result):
                        if(type_path == '/proc'):
                            all_failures += [Failure(code_filepath, idx + 1, f"'Global proc '/proc{rematch_result.group(4)}' found in a file restricted to type '{type_path}'")]
                        else:
                            all_failures += [Failure(code_filepath, idx + 1, f"'Definition for different type '{rematch_result.group(1)}' found in a file restricted to '{type_path}'")]
```

Reply:
> Furthermore, change the last group from `((/[\w]+)\(.*)?` to `((/[\w]+)\()?`(remove `.*`) as it does not appear that the contents of the proc's parameters alter the message.

We need to consume the entire string to force the non-greedy first group to collect everything it should, and we need to only catch trailing characters if we opened a proc.

Reply:
Updated the regex to also catch trailing comments on non-prod lines.

If you want to update the regex further, here are some test cases to run against it in command line:
```py3
>>> r.match("/proc/foo()   // TEST").groups()
('/proc', None, '/foo()')
>>> r.match("/proc/foo(   // TEST").groups()
('/proc', None, '/foo(')
>>> r.match("/some/path/proc/foo(   // TEST").groups()
('/some/path', '/proc', '/foo(')
>>> r.match("/some/path/proc // test").groups()
('/some/path', '/proc', None)
```

Reply:
This still trips up in some cases, this is why I wanted it to not include an end of line indicator, as there may be a bunch of gibberish after the proc name which is crazy to parse
![image](https://github.com/ParadiseSS13/Paradise/assets/91113370/2df779e7-2631-454f-923b-4ad833cf9af2)
![image](https://github.com/ParadiseSS13/Paradise/assets/91113370/12efbefe-acea-4f02-b1ff-ce2e3cb9b43b)
![image](https://github.com/ParadiseSS13/Paradise/assets/91113370/e08f4dba-71f0-463c-a4e1-42fc7f79a643)

Reply:
Updated the suggestion, covered your new cases.

```py3
>>> r = re.compile('^(/[\\w][\\w/]*?)(?:/proc)?(?:/([\\w]+)\\(.*?)?(?: */[/*].*)?$')
		       
>>> test_cases = ["/proc/foo()   // TEST", "/proc/foo(   // TEST", "/some/path/proc/foo(   // TEST", "/some/path/proc // test", "/mob/proc_holder/shit//proc/fuck(obj/mob) // lmaooo", "/mob/proc_holder/shit/proc/fuck(obj/mob) /* lmaooo", "/obj /* fuck"]
		       
>>> for t in test_cases:
	print("%s\n-->%r" % (t, r.match(t).groups()))

		       
/proc/foo()   // TEST
-->('/proc', 'foo')
/proc/foo(   // TEST
-->('/proc', 'foo')
/some/path/proc/foo(   // TEST
-->('/some/path', 'foo')
/some/path/proc // test
-->('/some/path', None)
/mob/proc_holder/shit//proc/fuck(obj/mob) // lmaooo
-->('/mob/proc_holder/shit', None)
/mob/proc_holder/shit/proc/fuck(obj/mob) /* lmaooo
-->('/mob/proc_holder/shit', 'fuck')
/obj /* fuck
-->('/obj', None)
```